### DISCUSSION TITLE
Trouble Running CPPPotential in the HPMC Module with GPU

### DISCUSSION THREAD

**Discussion by dchien79** [glotzerlab/hoomd-blue#1745] [Support] [answered]
Hello! I am trying to implement a user-defined potential in HOOMD while accelerating my HPMC simulations with the GPU. I previously installed the HOOMD GPU-enabled binary for CUDA 12.2, and confirmed that HOOMD is picking up the GPU.

I copied the tutorial script in the following webpage (https://hoomd-blue.readthedocs.io/en/v4.5.0/howto/cpppotential.html) to a python script called test.py and obtained the expected output. However, when I changed the line `sim = hoomd.Simulation(device=hoomd.device.CPU(), seed=1)` to `sim = hoomd.Simulation(device=hoomd.device.GPU(), seed=1)`, I was returned with the following error:

(hoomd2) [dchien@bhg0042 benchmark]$ python test.py
/gpfs/fs2/scratch/dchien/esc-hoomd/hoomd/benchmark/test.py:25: FutureWarning: CPPPotential is deprecated since 4.5.0. Use a hpmc.pair.Pair potential.
  square_well = hoomd.hpmc.pair.user.CPPPotential(r_cut=r_interaction,
/gpfs/fs2/scratch/dchien/esc-hoomd/hoomd/benchmark/test.py:34: FutureWarning: pair_potential is deprecated since 4.5.0. Use pair_potentials.
  mc.pair_potential = square_well
Traceback (most recent call last):
  File "/gpfs/fs2/scratch/dchien/esc-hoomd/hoomd/benchmark/test.py", line 42, in <module>
    sim.run(0)
  File "/software/miniconda3/4.12.0/envs/hoomd2/lib/python3.12/site-packages/hoomd/simulation.py", line 555, in run
    self.operations._schedule()
  File "/software/miniconda3/4.12.0/envs/hoomd2/lib/python3.12/site-packages/hoomd/operations.py", line 211, in _schedule
    self.integrator._attach(sim)
  File "/software/miniconda3/4.12.0/envs/hoomd2/lib/python3.12/site-packages/hoomd/operation.py", line 315, in _attach
    self._attach_hook()
  File "/software/miniconda3/4.12.0/envs/hoomd2/lib/python3.12/site-packages/hoomd/hpmc/integrate.py", line 457, in _attach_hook
    self._pair_potential._attach(self._simulation)
  File "/software/miniconda3/4.12.0/envs/hoomd2/lib/python3.12/site-packages/hoomd/operation.py", line 315, in _attach
    self._attach_hook()
  File "/software/miniconda3/4.12.0/envs/hoomd2/lib/python3.12/site-packages/hoomd/hpmc/pair/user.py", line 276, in _attach_hook
    self._cpp_obj = _jit.PatchEnergyJITGPU(
                    ^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Runtime compilation failed:
__cuda/cstdint_prelude.h(80): warning #47-D: incompatible redefinition of macro "UINT64_C" (declared at line 124 of nvccfeatures.h)
      #define UINT64_C(X) ((uint_least64_t)(X))
              ^

Remark: The warnings can be suppressed with "-diag-suppress <warning-number>"

../limits(218): error: identifier "__CHAR_BIT__" is undefined
      static _LIBCUDACXX_CONSTEXPR const int  digits = static_cast<int>(sizeof(type) * __CHAR_BIT__ - is_signed);
                                                                                       ^

../limits(319): error: identifier "__FLT_MANT_DIG__" is undefined
      static _LIBCUDACXX_CONSTEXPR const int  digits = __FLT_MANT_DIG__;
                                                       ^

../limits(320): error: identifier "__FLT_DIG__" is undefined
      static _LIBCUDACXX_CONSTEXPR const int  digits10 = __FLT_DIG__;
                                                         ^

../limits(328): error: identifier "__FLT_RADIX__" is undefined
      static _LIBCUDACXX_CONSTEXPR const int  radix = __FLT_RADIX__;
                                                      ^

../limits(332): error: identifier "__FLT_MIN_EXP__" is undefined
      static _LIBCUDACXX_CONSTEXPR const int  min_exponent = __FLT_MIN_EXP__;
                                                             ^

../limits(333): error: identifier "__FLT_MIN_10_EXP__" is undefined
      static _LIBCUDACXX_CONSTEXPR const int  min_exponent10 = __FLT_MIN_10_EXP__;
                                                               ^

../limits(334): error: identifier "__FLT_MAX_EXP__" is undefined
      static _LIBCUDACXX_CONSTEXPR const int  max_exponent = __FLT_MAX_EXP__;
                                                             ^

../limits(335): error: identifier "__FLT_MAX_10_EXP__" is undefined
      static _LIBCUDACXX_CONSTEXPR const int  max_exponent10 = __FLT_MAX_10_EXP__;
                                                               ^

../limits(322): error: identifier "__FLT_MIN__" is undefined
      _LIBCUDACXX_INLINE_VISIBILITY static _LIBCUDACXX_CONSTEXPR type min() _NOEXCEPT {return __FLT_MIN__;}
                                                                                              ^

../limits(323): error: identifier "__FLT_MAX__" is undefined
      _LIBCUDACXX_INLINE_VISIBILITY static _LIBCUDACXX_CONSTEXPR type max() _NOEXCEPT {return __FLT_MAX__;}
                                                                                              ^

../limits(329): error: identifier "__FLT_EPSILON__" is undefined
      _LIBCUDACXX_INLINE_VISIBILITY static _LIBCUDACXX_CONSTEXPR type epsilon() _NOEXCEPT {return __FLT_EPSILON__;}
                                                                                                  ^

../limits(343): error: identifier "__builtin_huge_valf" is undefined
      _LIBCUDACXX_INLINE_VISIBILITY static type infinity() _NOEXCEPT {return __builtin_huge_valf();}
                                                                             ^

../limits(344): error: identifier "__builtin_nanf" is undefined
      _LIBCUDACXX_INLINE_VISIBILITY static type quiet_NaN() _NOEXCEPT {return __builtin_nanf("");}
                                                                              ^

../limits(345): error: identifier "__builtin_nansf" is undefined
      _LIBCUDACXX_INLINE_VISIBILITY static type signaling_NaN() _NOEXCEPT {return __builtin_nansf("");}
                                                                                  ^

../limits(351): error: identifier "__FLT_DENORM_MIN__" is undefined
      _LIBCUDACXX_INLINE_VISIBILITY static _LIBCUDACXX_CONSTEXPR type denorm_min() _NOEXCEPT {return __FLT_DENORM_MIN__;}
                                                                                                     ^

../limits(371): error: identifier "__DBL_MANT_DIG__" is undefined
      static _LIBCUDACXX_CONSTEXPR const int  digits = __DBL_MANT_DIG__;
                                                       ^

../limits(372): error: identifier "__DBL_DIG__" is undefined
      static _LIBCUDACXX_CONSTEXPR const int  digits10 = __DBL_DIG__;
                                                         ^

../limits(380): error: identifier "__FLT_RADIX__" is undefined
      static _LIBCUDACXX_CONSTEXPR const int  radix = __FLT_RADIX__;
                                                      ^

../limits(384): error: identifier "__DBL_MIN_EXP__" is undefined
      static _LIBCUDACXX_CONSTEXPR const int  min_exponent = __DBL_MIN_EXP__;
                                                             ^

../limits(385): error: identifier "__DBL_MIN_10_EXP__" is undefined
      static _LIBCUDACXX_CONSTEXPR const int  min_exponent10 = __DBL_MIN_10_EXP__;
                                                               ^

../limits(386): error: identifier "__DBL_MAX_EXP__" is undefined
      static _LIBCUDACXX_CONSTEXPR const int  max_exponent = __DBL_MAX_EXP__;
                                                             ^

../limits(387): error: identifier "__DBL_MAX_10_EXP__" is undefined
      static _LIBCUDACXX_CONSTEXPR const int  max_exponent10 = __DBL_MAX_10_EXP__;
                                                               ^

../limits(374): error: identifier "__DBL_MIN__" is undefined
      _LIBCUDACXX_INLINE_VISIBILITY static _LIBCUDACXX_CONSTEXPR type min() _NOEXCEPT {return __DBL_MIN__;}
                                                                                              ^

../limits(375): error: identifier "__DBL_MAX__" is undefined
      _LIBCUDACXX_INLINE_VISIBILITY static _LIBCUDACXX_CONSTEXPR type max() _NOEXCEPT {return __DBL_MAX__;}
                                                                                              ^

../limits(381): error: identifier "__DBL_EPSILON__" is undefined
      _LIBCUDACXX_INLINE_VISIBILITY static _LIBCUDACXX_CONSTEXPR type epsilon() _NOEXCEPT {return __DBL_EPSILON__;}
                                                                                                  ^

../limits(395): error: identifier "__builtin_huge_val" is undefined
      _LIBCUDACXX_INLINE_VISIBILITY static type infinity() _NOEXCEPT {return __builtin_huge_val();}
                                                                             ^

../limits(396): error: identifier "__builtin_nan" is undefined
      _LIBCUDACXX_INLINE_VISIBILITY static type quiet_NaN() _NOEXCEPT {return __builtin_nan("");}
                                                                              ^

../limits(397): error: identifier "__builtin_nans" is undefined
      _LIBCUDACXX_INLINE_VISIBILITY static type signaling_NaN() _NOEXCEPT {return __builtin_nans("");}
                                                                                  ^

../limits(403): error: identifier "__DBL_DENORM_MIN__" is undefined
      _LIBCUDACXX_INLINE_VISIBILITY static _LIBCUDACXX_CONSTEXPR type denorm_min() _NOEXCEPT {return __DBL_DENORM_MIN__;}
                                                                                                     ^

29 errors detected in the compilation of "default_program".

I am wondering, then, if the CPPPotential function is supported when using the GPU as a device. Is there a way to fix this issue or is there a different function in HOOMD that can run user-defined potentials while still using the GPU? Let me know if so. Thank you!

**Comment by joaander** [comment, marked as answer]
In the newly released HOOMD-blue 4.6.0, you can use [`hpmc.pair.Step` and `Union` potentials](https://hoomd-blue.readthedocs.io/en/v4.6.0/module-hpmc-pair.html) without needing `CPPPotential`. The new potentials do not support the GPU, but that is for the best as the GPU implementation of `CPPPotential` was almost always much less efficient than the CPU. Furthermore, `hpmc.pair.Step` is much faster than an equivalent implementation with `CPPPotential` on the CPU.

I have not previously seen the specific error you report, but I have seen many like it. Fixing these usually requires many days of effort to replicate the error (matching compiler and CUDA environments) and then finding the right tweak to make to `jitify.hpp`.

As you can see in your output, `CPPPotential` is deprecated. I have no plans to further maintain it. In HOOMD-blue 4.6, `hpmc.pair` (excluding `pair.user`) covers almost all the cases I have seen people using `CPPPotential` for. In later 4.x releases, I plan to replace the external compiled potentials, provide a template component to help migrate custom `CPPPotential` cases, and update the tutorial. 5.0 will remove all `CPPPotential` classes.

**Reply by dchien79** [reply]
Okay, thank you for the assistance!
